src/components/TodoList.tsx

```diff
@@ -0,0 +1,152 @@
+import { useContext, useState } from 'react';
+import cn from 'classnames';
+import { TodosContext } from './TodoProvider';
+import * as postServise from '../api/todos';
+import { Errors } from '../types/Errors';
+
+export const TodoList: React.FC = () => {
+  const [isEditing, setIsEditing] = useState(false);
+  const [editId, setEditId] = useState<number | null>(null);
+  const {
+    todos,
+    setTodos,
+    setError,
+    removeTodo,
+    filteredTodo,
+  } = useContext(TodosContext);
+
+  const toggleTodo = async (todoId: number) => {
+    const completedTodo = todos.map(todo => (
+      todo.id === todoId
+        ? { ...todo, completed: !todo.completed }
+        : { ...todo }
+    ));
+    const currentTodo = completedTodo.find(complted => complted.id === todoId);
+
+    try {
+      await postServise.updateTodo({
+        todo: currentTodo,
+        todoId,
+      });
+    } catch {
+      setError(Errors.UNABLE_UPDATE);
+    }
+
+    setTodos(completedTodo);
+  };
+
+  const handleDoubleClick = (id: number) => {
+    setEditId(id);
+    setIsEditing(true);
+  };
+
+  const handleBlur = () => {
+    setIsEditing(false);
+  };
+
+  const handleKeyPress = async (
+    event: React.KeyboardEvent<HTMLInputElement>,
+    editingId: number,
+  ) => {
+    if (event.key === 'Enter') {
+      event.preventDefault();
+      setIsEditing(false);
+
+      const currentTodo = todos.find(todo => todo.id === editingId);
+
+      if (currentTodo) {
+        try {
+          const updatedTitle = event.currentTarget.value;
+          const updatedTodos = todos.map(todo => {
+            if (todo.id === editingId) {
+              return { ...todo, title: updatedTitle };
+            }
+
+            return todo;
+          });
+
+          await postServise.updateTodo({
+            todo: { ...currentTodo, title: updatedTitle },
+            todoId: editingId,
+          });
+
+          setTodos(updatedTodos);
+        } catch (updateError) {
+          setError(Errors.UNABLE_UPDATE);
+        }
+      }
+    }
+  };
+
+  const handleEditingChange = (
+    event: React.ChangeEvent<HTMLInputElement>,
+    editingId: number,
+  ) => {
+    const completedTodo = todos.map(todo => {
+      if (todo.id === editingId) {
+        return { ...todo, title: event.target.value };
+      }
+
+      return todo;
+    });
+
+    setTodos(completedTodo);
+  };
+
+  return (
+    <section className="todoapp__main" data-cy="TodoList">
+      {filteredTodo.map((todo) => (
+        <div
```

Comment:
I'd recommend you to move `TodoItem` to a separate component

it will simplify a lot of methods - they should accept entire todo object as an argument if needed